solr/core/src/java/org/apache/solr/handler/loader/JsonLoader.java

```diff
@@ -668,7 +682,40 @@ private boolean isChildDoc(SolrInputDocument extendedMap) {
       return extendedMap.containsKey(req.getSchema().getUniqueKeyField().getName());
     }
 
-    private SolrInputDocument generateExtendedValueMap(int ev) throws IOException {
+    private boolean entryIsChildDoc(Object val) {
+      if(val instanceof List) {
+        List listVal = (List) val;
+        if (listVal.size() == 0) return false;
+        return  listVal.get(0) instanceof Map;
+      }
+      return val instanceof Map;
+    }
+
+    private void safeAddValue(SolrInputDocument doc, String fieldName, Object value) {
```

Comment:
I'm confused why this is necessary.  What goes wrong if doc.addField(name,value) is called?  If something goes wrong and this is necessary, that explanation needs to be in the comments.

Reply:
I tried the same method, except because SolrInputDocument implements iterable, each document key is added, instead of the SolrInputDocument object. This chain of event occurs because [SolrInputDocument.addField](https://github.com/moshebla/lucene-solr/blob/3c51c1c414c6f7cb359e3ef442ca706e33cf3a7a/solr/solrj/src/java/org/apache/solr/common/SolrInputDocument.java#L84) calls [SolrInputField.addValue](https://github.com/moshebla/lucene-solr/blob/3c51c1c414c6f7cb359e3ef442ca706e33cf3a7a/solr/solrj/src/java/org/apache/solr/common/SolrInputField.java#L91).

Reply:
Okay I see.  Interesting -- there are clearly pros/cons to having SolrInputDocument implement core JDK abstractions.  It makes working with it easier, though now some things like what you see here happens automatically when we don't want it to.

There is no perfect solution here but I'd rather see SolrInputField.addValue do an instanceof check when it sees that `v` implements Iterable to guard against iterating the child doc.  This is much smaller/simpler than your safeAddValue, and it's something that won't be tied to Json or any other input format.  Even someone who wants to use SolrJ and thinks that they can just call set/addValue would be surprised it doesn't work unless we make the change at SolrInputField.
Also admittedly, this is a gotcha to the path of putting nested child docs in values, but not a big deal.

Reply:
So if I understood correctly, I also thought of checking to see if the addField value is of SolrDocumentBase, to prevent going through the iterator. I was afraid of backwards compatibility, but after some thought, until [SOLR-12361](https://issues.apache.org/jira/browse/SOLR-12361) childDocuments inside doc keys was not supported.